File: src/main/java/mine/leetcode/max_path_sum/MaxPathSum.py
'''
首先，想一个简化版(single path)，找从root到任意点得最大值。类似于maxDepth，每次加root.val而不再是+1

求单路的时候，如果root加左儿子单路或者右儿子单路最后的值都小于0，则返回0，意味着不要root开始的这个单路了
本题思路，divide & conquer
求最大路径和就等于下面三个值的最大值：
左子树的最大路径和
右子树最大路径和
左子树单路 + 右子树单路 + root.val
'''

class Solution(object):
    def maxPathSum(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        res, _ = self.helper(root)
        return res
        
    def helper(self, root):
        if not root:
            return -sys.maxint, 0
            
        left = self.helper(root.left)
        right = self.helper(root.right)
        singlePathSum = max(left[1] + root.val, right[1] + root.val, 0)
        maxPathSum = max(left[0], right[0], left[1] + right[1] + root.val)
        return maxPathSum, singlePathSum

File: src/main/java/mine/leetcode/max_path_sum/MaxPathSum.java
package mine.leetcode.max_path_sum;

import mine.leetcode.TreeNode;

/**
 * MaxPathSum <a href="www.example.com"></a>
 */
public class MaxPathSum {
    private int max = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        singlePathMax(root);
        return max;
    }

    private int singlePathMax(TreeNode r) {
        if (r == null)
            return 0;
        int left = singlePathMax(r.left);
        int right = singlePathMax(r.right);
        left = left > 0 ? left : 0;// 这两步很关键啊……何时以及如何废掉一个结果
        right = right > 0 ? right : 0;
        // int tmp = left > right ? left : right;
        // max = Math.max(max, Math.max(tmp, left + right + r.val));
        // return tmp + r.val;

        max = Math.max(max, left + right + r.val);
        return (left > right ? left : right) + r.val;
    }

    // int leftSingle, rightSingle;

    // private int helper(TreeNode r) {
    // int leftMax = helper(r.left), rightMax = helper(r.right);
    // leftSingle = single(r.left);
    // rightSingle = single(r.right);
    // int tmp = Math.max(leftSingle + rightSingle + r.val, Math.max(leftMax,
    // rightMax));
    // if (tmp > max)
    // max = tmp;
    // return tmp;
    // }

    // private int single(TreeNode r) {
    // return Math.max(leftSingle, rightSingle) + r.val;
    // }

    // private int maxPathSumFrom(TreeNode r) {
    // if (r == null)
    // return 0;
    // int left = singlePathSum(r.left);
    // int right = singlePathSum(r.right);
    // left = left > 0 ? left : 0;
    // right = right > 0 ? right : 0;

    // }

    // private int singlePathSum(TreeNode r) {
    // if(r==null)
    // return 0;
    // singlePathSum(r.left)
    // }

    public static void main(String[] args) {
        String[] test = { "[1,2,null,3,null,4,null,5]", "[1,2,3]", "[-3]", "[-10,9,20,null,null,15,7]" };
        for (String str : test) {
            TreeNode root = TreeNode.stringToTreeNode(str);
            MaxPathSum t = new MaxPathSum();
            int ret = t.maxPathSum(root);
            System.out.println(ret);
        }
    }
}